### DISCUSSION TITLE
Help me find await tx.wait() analog pls

### DISCUSSION THREAD

**Discussion by OoXooOx** [wevm/viem#960] [Question] [answered]
I start learn new viem+wagmi. 
I have project with below dependencies:
"dependencies": {
    "@emotion/react": "^11.11.1",
    "@emotion/styled": "^11.11.0",
    "@fontsource/roboto": "^5.0.5",
    "@mui/icons-material": "^5.14.1",
    "@mui/material": "^5.14.1",
    "@web3modal/ethereum": "^2.4.1",
    "@web3modal/react": "^2.4.1",
    "buffer": "^6.0.3",
    "env-cmd": "^10.1.0",
    "process": "^0.11.10",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-router-dom": "^6.7.0",
    "tss-react": "^4.3.4",
    "util": "^0.12.4",
    "viem": "^0.3.50",
    "wagmi": "^1.3.9"
  },

When it was old version of wagmi I use ethers and useContract hook for wait for confirming transaction and it's work perfectly.
Something like this. I was able invoke approve and stake through 1 button.

let allowance = Number(ethers.utils.formatEther(await TokenProvider?.allowance(address, STAKE_ADDRESS)))
            while (allowance < TokenToStake!) {
                const approve = await TokenSigner?.approve(STAKE_ADDRESS, ethers.utils.parseEther(TokenToStake!.toString()));
                setSpinner(true);
                await approve.wait(); // HERE WE WAITING 
                allowance = Number(ethers.utils.formatEther(await TokenProvider?.allowance(address, STAKE_ADDRESS)))
            }
            const tx = await StakeSigner?.stake(ethers.utils.parseEther(TokenToStake!.toString()), activeButton)
            setSpinner(true);
            await tx.wait();

But now we don't have useContract, we don't have ethers. So, what should I do? I need somehow wait until user give to my SC approve and only then invoke function in SC. How I can do this now?

**Comment by 0xLeonard25** [comment]
You can use waitForTransactionReceipt(), which is similar to tx.wait()
Reference: https://viem.sh/docs/actions/public/waitForTransactionReceipt.html

**Reply by OoXooOx** [reply]
So, where from I can fetch tx.hash for waitForTransactionReceipt()? I try it and I have undefined field. Do you have code snippet?

**Reply by 0xLeonard25** [reply]
You can get the txHash from the return value of writeContract()
Reference: https://viem.sh/docs/contract/writeContract.html

**Comment by OoXooOx** [comment]
It's hard but I receive hash when I use viem directly.  
const approve = await walletClient.writeContract({
                ...USDT_ContractConfig,
                functionName: 'approve',
                args: [NFT_ContractConfig.address, amount],
                account: address as `0x${string}`
            })
            console.log(approve); // tx.hash here/

But it's work only with  transport: custom(window.ethereum):
export const walletClient = createWalletClient({
  chain: mainnet,
  transport: custom(window.ethereum)
})

How I can fetch transport from web3modal connection?

**Reply by jxom** [reply]
Here is an example of a vanilla web3modal provider: https://github.com/WalletConnect/web3modal-examples/blob/main/walletconnect-modal-ethereum-provider-html/main.js

All you have to do to make Web3Modal work with viem is put the `ethereumProvider` in your `custom` transport instead of `window.ethereum`.

**Comment by OoXooOx** [comment]
I need this wallets. This window I receive when use web3modal. Here Metamask, Trust, MEW etc.
![MM](https://github.com/wagmi-dev/viem/assets/107791774/2588a783-af68-4747-97ad-f3e3440c67cd)
If I use code which you provide, I receive bellow pop up window. There is no Metamask.
![WC](https://github.com/wagmi-dev/viem/assets/107791774/bf920f7b-2341-4465-ac05-929eb3bc0fc0)
And in this https://github.com/WalletConnect/web3modal-examples/blob/main/walletconnect-modal-ethereum-provider-html/main.js code we import { EthereumProvider } from "@walletconnect/ethereum-provider";  then we need connect it   ethereumProvider = await EthereumProvider.init({   But I already have provider from web3modal when I connect my wallet through it, right? Why I must connect it again?  I try paste in transport                     import { EthereumClient } from '@web3modal/ethereum'         const ethereumClient = new EthereumClient(config, chains) but it doesn't work. I don't understand how fix it....

**Comment by OoXooOx** [comment, marked as answer]
I can't resolve question with transport, but ...
ok, I understand that I completely confused. useEffect helped me in my case:
const { write: writeMint, data: mintData, error: mintError, isLoading: mintIsLoading, isError: mintIsError } = useContractWrite({
        ...NFT_ContractConfig,
        functionName: 'mint',
    })
const { write: writeApprove, data: approveData, error: approveError, isLoading: approveIsLoading, isError: approveIsError }
        = useContractWrite({
            ...USDT_ContractConfig,
            functionName: 'approve',
            args: [NFT_ContractConfig.address, 2000000000],
        })
const { isSuccess: approveIsSuccess } = useWaitForTransaction({ hash: approveData?.hash, })
useEffect(() => {
        approveIsSuccess &&
            //  && trigger
            writeMint({
                args: [2],
            });
        // trigger=false;
    }, [approveIsSuccess])

const buyNFT = async ()=>{
           writeApprove();
}
